File: solutions/2026-03-04/cpp/2026-03-04-special-positions-in-a-binary-matrix.cpp
// Problem: Special Positions in a Binary Matrix
// Link: https://leetcode.com/problems/special-positions-in-a-binary-matrix/
//
// Approach:
// We need to count the number of cells (i, j) where mat[i][j] is 1 and
// all other elements in the same row 'i' and same column 'j' are 0.
//
// A straightforward approach is to iterate through each cell of the matrix.
// If a cell mat[i][j] contains a 1, we then check its entire row and column.
// To efficiently check rows and columns, we can precompute the sum of each row
// and the sum of each column.
//
// For each cell (i, j):
// 1. If mat[i][j] == 1:
//    a. Check if the sum of row 'i' is exactly 1.
//    b. Check if the sum of column 'j' is exactly 1.
//    c. If both conditions are true, then (i, j) is a special position.
//
// Time Complexity:
// O(m * n), where 'm' is the number of rows and 'n' is the number of columns.
// We iterate through the matrix once to compute row sums and column sums,
// which takes O(m*n). Then, we iterate through the matrix again to check
// each cell and its corresponding precomputed sums, which also takes O(m*n).
//
// Space Complexity:
// O(m + n), for storing the sums of each row and each column.
// We use two arrays (or vectors): one of size 'm' for row sums and one of size 'n' for column sums.

#include <vector>
#include <numeric> // For std::accumulate

class Solution {
public:
    int numSpecial(std::vector<std::vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();

        // Vectors to store the sum of each row and each column.
        // row_sums[i] will store the sum of elements in row i.
        // col_sums[j] will store the sum of elements in column j.
        std::vector<int> row_sums(m, 0);
        std::vector<int> col_sums(n, 0);

        // First pass: Compute the sum of each row and each column.
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                row_sums[i] += mat[i][j];
                col_sums[j] += mat[i][j];
            }
        }

        int special_positions_count = 0;

        // Second pass: Iterate through the matrix to find special positions.
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                // A position (i, j) is special if:
                // 1. The element at mat[i][j] is 1.
                // 2. The sum of row 'i' is exactly 1 (meaning only mat[i][j] is 1 in this row).
                // 3. The sum of column 'j' is exactly 1 (meaning only mat[i][j] is 1 in this column).
                if (mat[i][j] == 1 && row_sums[i] == 1 && col_sums[j] == 1) {
                    special_positions_count++;
                }
            }
        }

        return special_positions_count;
    }
};

File: solutions/2026-03-04/javascript/2026-03-04-special-positions-in-a-binary-matrix.js
// Problem Summary: Count cells in a binary matrix that have a 1 and are the only 1 in their row and column.
// Problem Link: https://leetcode.com/problems/special-positions-in-a-binary-matrix/
// Approach:
// 1. Precompute the sum of 1s for each row and each column.
// 2. Iterate through the matrix. If a cell contains a 1, check if its corresponding row sum and column sum are both 1.
// 3. If both conditions are met, increment the count of special positions.
// Time Complexity: O(m * n), where m is the number of rows and n is the number of columns. We iterate through the matrix once to calculate sums and once to count special positions.
// Space Complexity: O(m + n), for storing the row sums and column sums.
const numSpecial = function(mat) {
    const m = mat.length;
    const n = mat[0].length;

    // Initialize arrays to store the sum of 1s in each row and column.
    const rowSums = new Array(m).fill(0);
    const colSums = new Array(n).fill(0);

    // Calculate the sum of 1s for each row and column.
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (mat[i][j] === 1) {
                rowSums[i]++;
                colSums[j]++;
            }
        }
    }

    let specialCount = 0;

    // Iterate through the matrix again to find special positions.
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            // A position (i, j) is special if it contains a 1 and is the only 1 in its row and column.
            if (mat[i][j] === 1 && rowSums[i] === 1 && colSums[j] === 1) {
                specialCount++;
            }
        }
    }

    // Return the total count of special positions.
    return specialCount;
};
